File: solutions/cpp/Gassing_up.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::pair;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
size_t FindStartCity(const vector<int>& G, const vector<int>& D) {
  int carry = 0;
  pair<size_t, int> city_carry_pair(0, 0);
  for (size_t i = 1; i < G.size(); ++i) {
    carry += G[i - 1] - D[i - 1];
    if (carry < city_carry_pair.second) {
      city_carry_pair = {i, carry};
    }
  }
  return city_carry_pair.first;
}
// @exclude

void CheckAns(const vector<int>& G, const vector<int>& D, size_t c) {
  size_t s = c;
  int gas = 0;
  do {
    gas += G[s] - D[s];
    assert(gas >= 0);
    s = (s + 1) % G.size();
  } while (s != c);
}

void SmallTest() {
  // Example in the book.
  vector<int> G = {20, 15, 15, 15, 35, 25, 30, 15, 65, 45, 10, 45, 25};
  vector<int> D = {15, 20, 50, 15, 15, 30, 20, 55, 20, 50, 10, 15, 15};
  size_t ans = FindStartCity(G, D);
  assert(ans == 8);
  CheckAns(G, D, ans);
}

int main(int argc, char* argv[]) {
  SmallTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> G, D;
    int sum = 0;
    for (int i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(1, 200);
      int x = dis(gen);
      sum += x;
      G.emplace_back(x);
    }
    sum -= n;
    for (int i = 0; i < n; ++i) {
      int x = 0;
      if (sum) {
        uniform_int_distribution<int> dis(1, sum);
        x = dis(gen);
      }
      D.emplace_back(x + 1);
      sum -= x;
    }
    D.back() += sum;
    /*
    for (int i = 0; i < n; ++i) {
      cout << G[i] << ' ';
    }
    cout << endl;
    for (int i = 0; i < n; ++i) {
      cout << D[i] << ' ';
    }
    cout << endl;
    */
    size_t c = FindStartCity(G, D);
    cout << "start city = " << c << endl;
    CheckAns(G, D, c);
  }
  return 0;
}

File: solutions/python/gassing_up.py
import sys
import random
import collections

# @include
MPG = 20


# gallons[i] is the amount of gas in city i, and distances[i] is the
# distance city i to the next city.
def find_ample_city(gallons, distances):
    remaining_gallons = 0
    CityAndRemainingGas = collections.namedtuple('CityAndRemainingGas',
                                                 ('city', 'remaining_gallons'))
    city_remaining_gallons_pair = CityAndRemainingGas(0, 0)
    num_cities = len(gallons)
    for i in range(1, num_cities):
        remaining_gallons += gallons[i - 1] - distances[i - 1] // MPG
        if remaining_gallons < city_remaining_gallons_pair.remaining_gallons:
            city_remaining_gallons_pair = CityAndRemainingGas(i,
                                                              remaining_gallons)
    return city_remaining_gallons_pair.city
# @exclude


def check_ans(gallons, distances, c):
    s = c
    gas = 0
    while True:
        gas += gallons[s] - distances[s] // MPG
        assert gas >= 0
        s = (s + 1) % len(gallons)
        if s == c:
            break


def small_test():
    # Example in the book.
    gallons = (20, 15, 15, 15, 35, 25, 30, 15, 65, 45, 10, 45, 25)
    distances = (15 * MPG, 20 * MPG, 50 * MPG, 15 * MPG, 15 * MPG, 30 * MPG,
                 20 * MPG, 55 * MPG, 20 * MPG, 50 * MPG, 10 * MPG, 15 * MPG,
                 15 * MPG)
    ans = find_ample_city(gallons, distances)
    assert ans == 8
    check_ans(gallons, distances, ans)


def main():
    small_test()
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10000)
        gallons = []
        distances = []
        sum_ = 0
        for i in range(n):
            x = random.randint(1, 200)
            sum_ += x
            gallons.append(x)
        sum_ -= n
        for i in range(n):
            x = 0
            if sum_:
                x = random.randint(1, sum_)
            distances.append(x + 1)
            sum_ -= x
        distances[-1] += sum_
        c = find_ample_city(gallons, distances)
        print('start city =', c)
        check_ans(gallons, distances, c)


if __name__ == '__main__':
    main()
